File: sources/telegram/h8_payment.py
"""H8 payment / vertical gate. Keep in sync with internal/filter/h8_payment.go."""

from __future__ import annotations

import re

from .pain import message_has_tracker_pain
from .prefilter import has_buyer_question_signal

_FOP_TOV_RE = re.compile(
    r"(?i)(?:\bфоп\b|\bтов\b|\bfop\b|\btov\b|гривн|grivna|\buah\b|privatbank|monobank)"
)
_INFOBIZ_RE = re.compile(
    r"(?i)(?:инфобиз|infobiz|white\s+goods|курсы?\s+по\s+арбитраж|course\s+seller|"
    r"how\s+to\s+buy\s+domain|купить\s+домен)"
)
_CRYPTO_PAYOUT_RE = re.compile(
    r"(?i)(?:\busdt\b|\btrc20\b|\berc20\b|crypto\s+payout|crypto\s+settlement|"
    r"weekly\s+usdt|daily\s+usdt)"
)
_BUYER_VOICE_RE = re.compile(
    r"(?i)(?:voluum|keitaro|binom|redtrack|postback|tracker|usdt|trc20|clickid|"
    r"self-hosted|502|nginx|mysql)"
)


def reject_h8_payment_vertical(text: str, title: str = "") -> tuple[bool, str]:
    combined = f"{title} {text}".strip().lower()
    if not combined:
        return False, ""
    if _FOP_TOV_RE.search(combined) and not _BUYER_VOICE_RE.search(combined):
        return True, "h8 fop/tov/grivna"
    if _INFOBIZ_RE.search(combined) and not message_has_tracker_pain(text):
        return True, "h8 infobiz/off-vertical"
    return False, ""


def reject_crypto_payout_only(text: str) -> tuple[bool, str]:
    body = (text or "").strip()
    if not body:
        return False, ""
    if not _CRYPTO_PAYOUT_RE.search(body):
        return False, ""
    if message_has_tracker_pain(body):
        return False, ""
    if has_buyer_question_signal(body):
        return False, ""
    return True, "h8 crypto payout without pain"

File: internal/filter/h8_payment.go
package filter

import (
	"regexp"
	"strings"

	"github.com/bidshard/parser/internal/validate"
)

var (
	// Cyrillic terms omit \b: Go RE2 word boundaries are ASCII-only.
	h8FOPTOVRe = regexp.MustCompile(
		`(?i)(?:фоп|тов|гривн|grivna|\bfop\b|\btov\b|\buah\b|privatbank|monobank)`,
	)
	h8InfobizRe = regexp.MustCompile(
		`(?i)(?:инфобиз|infobiz|white\s+goods|курсы?\s+по\s+арбитраж|course\s+seller|` +
			`how\s+to\s+buy\s+domain|купить\s+домен)`,
	)
	h8CryptoPayoutRe = regexp.MustCompile(
		`(?i)(?:\busdt\b|\btrc20\b|\berc20\b|crypto\s+payout|crypto\s+settlement|weekly\s+usdt|daily\s+usdt)`,
	)
)

// RejectH8PaymentVertical drops local payment rails and off-vertical noise (H8/M11).
func RejectH8PaymentVertical(text, title string) (bool, string) {
	combined := strings.ToLower(strings.TrimSpace(title + " " + text))
	if combined == "" {
		return false, ""
	}
	if h8FOPTOVRe.MatchString(combined) && !hasH8BuyerVoice(combined) {
		return true, "h8 fop/tov/grivna"
	}
	if h8InfobizRe.MatchString(combined) && !validate.HasTrackerPainMessage(text) {
		return true, "h8 infobiz/off-vertical"
	}
	return false, ""
}

// RejectCryptoPayoutOnly rejects USDT/crypto mention without tracker/operational pain (H8).
func RejectCryptoPayoutOnly(text string) (bool, string) {
	body := strings.TrimSpace(text)
	if body == "" {
		return false, ""
	}
	lower := strings.ToLower(body)
	if !h8CryptoPayoutRe.MatchString(lower) {
		return false, ""
	}
	if validate.HasTrackerPainMessage(body) {
		return false, ""
	}
	if validate.HasCommercialPainIntent(body) || validate.HasBuyerQuestionPattern(body) {
		return false, ""
	}
	return true, "h8 crypto payout without pain"
}

func hasH8BuyerVoice(lower string) bool {
	for _, hint := range []string{
		"voluum", "keitaro", "binom", "redtrack", "postback", "tracker", "usdt",
		"trc20", "clickid", "self-hosted", "502", "nginx", "mysql",
	} {
		if strings.Contains(lower, hint) {
			return true
		}
	}
	return false
}
